homeassistant/components/dlink/switch.py

```diff
@@ -54,46 +60,68 @@ def setup_platform(
     discovery_info: DiscoveryInfoType | None = None,
 ) -> None:
     """Set up a D-Link Smart Plug."""
-
-    host = config[CONF_HOST]
-    username = config[CONF_USERNAME]
-    password = config[CONF_PASSWORD]
-    use_legacy_protocol = config[CONF_USE_LEGACY_PROTOCOL]
-    name = config[CONF_NAME]
-
-    smartplug = SmartPlug(host, password, username, use_legacy_protocol)
-    data = SmartPlugData(smartplug)
-
-    add_entities([SmartPlugSwitch(hass, data, name)], True)
-
-
-class SmartPlugSwitch(SwitchEntity):
+    async_create_issue(
+        hass,
+        DOMAIN,
+        "deprecated_yaml",
+        breaks_in_ha_version="2023.3.0",
+        is_fixable=False,
+        severity=IssueSeverity.WARNING,
+        translation_key="deprecated_yaml",
+    )
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={"source": SOURCE_IMPORT},
+            data=config,
+        )
+    )
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the D-Link Power Plug switch."""
+    async_add_entities(
+        [
+            SmartPlugSwitch(
+                hass,
+                entry,
+                SmartPlugData(hass.data[DOMAIN][entry.entry_id]),
```

Comment:
It would be a bit cleaner to store `SmartPlugData` in `async_setup_entry` instead and use it here.

https://github.com/home-assistant/core/pull/84927/files#r1064254925